### PR TITLE
Open with... improvement

### PR DESCRIPTION
**Pull request by hius07**
The purpose is to open non-document files by a tap in file browser and file search results.
Currently we have 4 auxiliary (non-document) providers: Image viewer, Text viewer, Text editor plugin, Archive viewer plugin.

Changes:
(1) **DocumentRegistry**
-a plugin can register itself as an auxiliary provider
-deduplicated code and run

(2) **Open with... dialog**
-associate a file type with an auxiliary provider
-view the list of providers associated with a file type
-reset the provider associated with a single file

Unchanged:
A single file cannot be associated with an auxiliary provider.
Document providers' standard behaviour.

(3) **Archive viewer**
-view images in an archive as a multi-page image (useful to view cbz content)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11056)

### CODE REVIEW

File: frontend/apps/filemanager/filemanagercollection.lua [hunk 1]
```diff
@@ -161,7 +161,7 @@ function FileManagerCollection:showCollDialog()
                     path = G_reader_settings:readSetting("home_dir"),
                     select_directory = false,
                     file_filter = function(file)
-                        return DocumentRegistry:getProviders(file) ~= nil
+                        return DocumentRegistry:hasProvider(file)
```
Comment (Frenzie) [inline frontend/apps/filemanager/filemanagercollection.lua:164]: Is the behavior the same for normally hidden files with a custom assigned engine and similar edge cases?

**Review by Frenzie — APPROVED** [review]
lgtm in principle, not able to test atm though

File: frontend/apps/filemanager/filemanagercollection.lua (hunk 1 above)
Comment (hius07) [reply frontend/apps/filemanager/filemanagercollection.lua:164]: Please give an example what do you mean.
We add to favorites only documents (ie they have a provider).

Comment (Frenzie) [reply frontend/apps/filemanager/filemanagercollection.lua:164]: I mainly meant I assumed it was a line I wrote a few years back, but I see now that it's in something you wrote. ;-)

**Comment by hius07** [comment]
Aligned infomessage with type associates.

**Comment by hius07** [comment]
Thank you, I understand. I'll change the architecture.

File: frontend/document/documentregistry.lua [hunk 2]
```diff
@@ -4,12 +4,12 @@ This is a registry for document providers
 
 local DocSettings = require("docsettings")
 local logger = require("logger")
-local lfs = require("libs/libkoreader-lfs")
 local util = require("util")
 
 local DocumentRegistry = {
     registry = {},
     providers = {},
+    providers_hash = {}, -- hash table of registered providers { provider_key = provider }
```
Comment (poire-z) [inline frontend/document/documentregistry.lua:12]: `providers_hash` sounds like it would be a hash value (a string or a number).
It is a Lua hash (not an array) table, ok, but we don't advertize that fact in most of our names :)
So, may be `known_providers` or  `provider_names` ?

**Review by poire-z — COMMENTED** [review]
Not a thorough review (this provider business is a bit too much for me :)), but nothing jumps out. Trusting you.
Just a nitpick.

**Comment by hius07** [comment]
@pazos, regarding your https://github.com/koreader/koreader/pull/9615 and https://github.com/koreader/koreader/issues/11001, I do not follow your way:
-because such handlers as Text viewer or Text editor can be used for any file types,
-and to avoid far route via showReader, then detect non-document, close Reader, reopen FM and handle the file.

I think your non-document handlers (plugins?) can be registered as auxiliary provider (see Text editor and Archive viewer here). Then any packet/non-document extension can be associated (manually via Open with... dialog or by the plugin itself) with your aux provider. It will be opened just by a tap in the file browser and in the file search results.

**Comment by pazos** [comment]
> @pazos, regarding your #9615 and #11001, I do not follow your way: -because such handlers as Text viewer or Text editor can be used for any file types, -and to avoid far route via "showReader, then detect non-document, close Reader, reopen FM and finally handle the file".
> 
> I think your non-document handlers (plugins?) can be registered as auxiliary provider (see Text editor and Archive viewer here). Then any packet/non-document extension can be associated (manually in Open with... dialog or automatically by the plugin itself) with your aux provider. It will be opened just by a tap in the file browser and in the file search results. No changes to the core would be required (except your icons in the coverbrowser, but that's another story).

Thanks for the hints. Feel free to do it your way. Mine was a quick attempt. Anyhow, I won't probably have time for that stuff until I'm retired :/

Will close those, while I'm at it :)

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 9 of 14 files at r1, 1 of 2 files at r2, 4 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11056)* status: all files reviewed, 6 unresolved discussions (waiting on @hius07)

___
*[`frontend/apps/filemanager/filemanagerfilesearcher.lua` line 322 at r4](https://reviewable.io/reviews/koreader/koreader/11056#-NiB0Q-EDtMtlxl-Type:-NiB0Q-EDtMtlxl-Typf:b-4mi8ec) ([raw file](https://github.com/koreader/koreader/blob/0b6cd1133e7ca4a1aa4cb53fbf4ec23aa3822435/frontend/apps/filemanager/filemanagerfilesearcher.lua#L322)):*
> ```Lua
>                 close_dialog_callback()
>                 local FileManager = require("apps/filemanager/filemanager")
>                 FileManager.openFile(self.ui, file, nil, self.close_callback)
> ```

Unfamiliar with all this, but I don't follow why `close_callback` is now limited to running when opening via auxiliary  providers?



_Code quote:_
```Lua
                close_dialog_callback()
                local FileManager = require("apps/filemanager/filemanager")
                FileManager.openFile(self.ui, file, nil, self.close_callback)
```
___
*[`frontend/document/documentregistry.lua` line 285 at r4](https://reviewable.io/reviews/koreader/koreader/11056#-NiB-18g5N_YiQM-Mp5F:-NiB-18g5N_YiQM-Mp5G:bhv9fym) ([raw file](https://github.com/koreader/koreader/blob/0b6cd1133e7ca4a1aa4cb53fbf4ec23aa3822435/frontend/document/documentregistry.lua#L285)):*
> ```Lua
> require("document/djvudocument"):register(DocumentRegistry)
> require("document/picdocument"):register(DocumentRegistry)
> -- auxuliary built-in
> ```

nit: `auxiliary` ;)
___
*[`frontend/ui/widget/imageviewer.lua` line 863 at r4](https://reviewable.io/reviews/koreader/koreader/11056#-NiB0zC59jRJ-nRDpK3a:-NiB0zC59jRJ-nRDpK3b:baxor90) ([raw file](https://github.com/koreader/koreader/blob/0b6cd1133e7ca4a1aa4cb53fbf4ec23aa3822435/frontend/ui/widget/imageviewer.lua#L863)):*
> ```Lua
>             return registry:isImageFile(file)
>         end,
>         callback = function(file)
> ```

You should be able to use a direct assignment here, e.g., `callback = ImageViewer.openFile,`, the signatures are identical, and the closure doesn't do anything fancy at all ;).
___
*[`frontend/ui/widget/textviewer.lua` line 544 at r4](https://reviewable.io/reviews/koreader/koreader/11056#-NiB1mLYBLzIZ_Kii1I4:-NiB1mLYBLzIZ_Kii1I5:b-vptpdn) ([raw file](https://github.com/koreader/koreader/blob/0b6cd1133e7ca4a1aa4cb53fbf4ec23aa3822435/frontend/ui/widget/textviewer.lua#L544)):*
> ```Lua
>         callback = function(file)
>             TextViewer.openFile(file)
>         end,
> ```

Ditto ;).



_Code quote:_
```Lua
        callback = function(file)
            TextViewer.openFile(file)
        end,
```

**Comment by NiLuJe** [comment]
I also wouldn't mind clearer comments about the fact that none of these aux stuff actually implement the Document* API, and we're just shoe-horning those in `DocumentRegistry`, which requires more careful consideration in how you actually deal with them to do something useful (e.g., in `FileManager.openFile`)

**Comment by hius07** [comment]
> why `close_callback` is now limited to running when opening via auxiliary providers?

We are in the search results list.
I want to view a file (say, in the Text viewer), keep the list opened and then view the next found file, so on.
When I open a file in Reader, the list should be closed.

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/08e0af69db3697e6f07b0da4c4d3223c6acb5b46/frontend/document/documentregistry.lua#L49-L52

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/4027288cf11dc1d41eac88a1764a533e6f174596/frontend/document/documentregistry.lua#L49-L54

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11056)* status: all files reviewed, 4 unresolved discussions (waiting on @hius07)

File: frontend/ui/widget/infomessage.lua [hunk 3]
```diff
@@ -77,6 +78,10 @@ local InfoMessage = InputContainer:extend{
 }
 
 function InfoMessage:init()
+    if not self.font_adjusted then
+        self.face = Font:getFace(self.monospace_font and "infont" or "infofont")
```
Comment (poire-z) [inline frontend/ui/widget/infomessage.lua:82]: Or just `if not self.face then self.face = ...` ?
(Will allow passing a `face=` when we want something other of infont/infofont?)
